File: src/graphs/python/prim/src/prim.py
"""Prim's algorithm for Minimum Spanning Tree (MST)."""

from __future__ import annotations

from heapq import heappush, heappop

from graph import Graph, Vertex


def prim(graph: Graph) -> float:
    """
    Find the minimum spanning tree weight using Prim's algorithm.

    Uses a priority queue (min-heap) for efficient O((V+E) log V) performance.

    Args:
        graph: The graph to find the MST for.

    Returns:
        The total weight of the minimum spanning tree.
    """
    visited: dict[Vertex, bool] = {}

    vertices = graph.vertices()

    for vertex in vertices:
        visited[vertex] = False

    # Pick first vertex as starting point
    vertex = vertices[0]
    visited[vertex] = True

    unvisited: list[tuple[float, Vertex]] = [
        (edge.distance, edge.destination) for edge in graph.edges_from_vertex(vertex)
    ]

    result = 0.0

    while unvisited:
        distance, vertex = heappop(unvisited)

        if not visited[vertex]:
            visited[vertex] = True
            result += distance
            for edge in graph.edges_from_vertex(vertex):
                if not visited[edge.destination]:
                    heappush(unvisited, (edge.distance, edge.destination))

    return result

File: src/graphs/cpp/prim/src/prim.cpp
#include "prim.h"
#include <queue>
#include <unordered_map>
#include <vector>

// Prim's algorithm using a priority queue for efficient minimum edge selection
template <class T> int prim(const Graph<T> &graph) {
  if (graph.empty())
    return 0;

  auto vertices = graph.vertices();
  if (vertices.empty())
    return 0;

  std::unordered_map<Vertex<T>, bool, HashFunction<T>> inMST;

  for (const auto &vertex : vertices)
    inMST[vertex] = false;

  // Start from the first vertex
  Vertex<T> source = vertices.front();
  inMST[source] = true;

  // Min-heap of edges ordered by weight
  auto cmp = [](const Edge<T> &a, const Edge<T> &b) {
    return a.distance > b.distance;
  };

  std::priority_queue<Edge<T>, std::vector<Edge<T>>, decltype(cmp)> minHeap(
      cmp);

  // Add all edges from source to the heap
  for (const auto &edge : graph.edgesFromVertex(source))
    minHeap.push(edge);

  int mstWeight = 0;
  size_t verticesInMST = 1;
  const size_t totalVertices = vertices.size();

  while (!minHeap.empty() && verticesInMST < totalVertices) {
    Edge<T> minEdge = minHeap.top();
    minHeap.pop();

    const auto &v = minEdge.destination;

    // Skip if vertex is already in MST
    if (inMST[v])
      continue;

    // Add vertex to MST
    inMST[v] = true;
    mstWeight += minEdge.distance;
    verticesInMST++;

    // Add all edges from newly added vertex
    for (const auto &edge : graph.edgesFromVertex(v)) {
      if (!inMST[edge.destination])
        minHeap.push(edge);
    }
  }

  return mstWeight;
}

template int prim<int>(const Graph<int> &graph);
template int prim<char>(const Graph<char> &graph);
template int prim<float>(const Graph<float> &graph);
template int prim<double>(const Graph<double> &graph);
